File: 백준/레벨Gold IV/1043. 거짓말/거짓말.cc
#include <bits/stdc++.h>

using namespace std;

vector<int> parent;

int find(int x) {
    vector<int> nodes;
    while (parent[x] != x) {
        nodes.push_back(x);
        x = parent[x];
    }
    
    for (int node : nodes) {
        parent[node] = x;
    }
    return x;
}

void unionSet(int a, int b) {
    int rootA = find(a);
    int rootB = find(b);
    
    if (rootA > rootB) parent[rootA] = rootB;
    else parent[rootB] = rootA;
}

void disjointSet() {
    // 부모노드 초기화
    int N, M;
    cin >> N >> M;

    parent.resize(N + 1);
    for (int i = 0; i <= N; i++) parent[i] = i;
    
    // 진실노드 초기화
    int truthSize;
    cin >> truthSize;

    if (truthSize == 0) {
        cout << M;
        return;
    }
    
    vector<int> truth(truthSize);
    for (int i = 0; i < truth.size(); i++) cin >> truth[i];
    
    // 파티노드 초기화
    vector<vector<int>> parties;
    for (int i = 0; i < M; i++) {
        int partySize;
        cin >> partySize;

        vector<int> party(partySize);
        for (int j = 0; j < party.size(); j++) cin >> party[j];
        parties.push_back(party);
        
        for (int j = 0; j < party.size() - 1; j++) {
            unionSet(party[j], party[j + 1]);
        }
    }
    
    // 진실노드로 병합
    int root = find(truth[0]);
    for (int i = 1; i < truth.size(); i++) {
        unionSet(truth[i], root);
    }
    
    // 파티 처리
    int count = M;
    for (const auto& party : parties) {
        if (find(party[0]) == find(root)) count--;
    }
    
    cout << count;
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    disjointSet();
}

File: 백준/레벨Gold IV/1043. 거짓말/거짓말.java
import java.io.*;
import java.util.*;

class Solution {
    int[] parent;

    public void disjointSet() throws IOException {
        // 부모노드 초기화
        int N = nextInt(), M = nextInt();
        parent = new int[N + 1];
        for (int i = 1; i <= N; i++) parent[i] = i;

        // 진실노드 초기화
        int truthSize = nextInt();
        if (truthSize == 0) {
            System.out.println(M);
            return;
        }

        int[] truth = new int[truthSize];
        for (int i = 0; i < truth.length; i++) truth[i] = nextInt();

        // 파티노드 초기화
        List<int[]> parties = new ArrayList<>();
        for (int i = 0; i < M; i++) {
            int partySize = nextInt();
            
            int[] party = new int[partySize];
            for (int j = 0; j < party.length; j++) party[j] = nextInt();
            parties.add(party);

            for (int j = 0; j < party.length - 1; j++) union(party[j], party[j + 1]);
        }

        // 진실노드로 병합
        int root = find(truth[0]);
        for (int i = 1; i < truth.length; i++) union(truth[i], root);

        // 파티 처리
        int count = M;
        for (int[] party : parties) {
            if (find(party[0]) == find(root)) count--;
        }

        System.out.println(count);
    }

    int find(int x) {
        List<Integer> nodes = new ArrayList<>();
        while (parent[x] != x) {
            nodes.add(x);
            x = parent[x];
        }

        for (int node : nodes) parent[node] = x;
        return x;
    }
    
    void union(int a, int b) {
        int root_a = find(a);
        int root_b = find(b);

        if (root_a > root_b) parent[root_a] = root_b;
        else parent[root_b] = root_a;
    }    

    int nextInt() throws IOException {
        int n = 0;
        int c;

        while ((c = System.in.read()) <= 32); // 탭 2, 개행 10, 공백 32

        do {
            n = n * 10 + (c - '0'); // 한자리씩 추가
        } while ((c = System.in.read()) > 32); // 구분자 : 공백

        return n;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        new Solution().disjointSet();
    }
}